### DISCUSSION TITLE
Developing a serverless (JS only) Mastodon front-end

### DISCUSSION THREAD

**Discussion by abuseofnotation** [mastodon/mastodon#27754] [General]
Last week I hacked together a [Mastodon feed aggregator](abuseofnotation.github.io/fediscope/) and began wondering the following: is it possible to turn it into a serverless (JS only) Mastodon front-end?

I think that this will be beneficial to the community, as it would allow for people to very easy write Mastodon web front-ends.

But currently the answer seems to be "No".
One reason is that to log the user in, I have to create applications in each instance, store their ID somewhere etc. 

So,is it possible to have a simplified OAuth flow where I just click Login and I am redirected directly to the front-end.

Also, I am wondering whether the CORS settings would allow such a thing?

**Comment by pkoepke** [comment]
I think Semaphore does what you're describing: https://github.com/NickColley/semaphore

**Comment by pkreissel** [comment]
I don't know exactly what you mean. Apps like elk and phanpy already are 100% Javascript Frontends. Also you can check out my Repos, I wrote a Algorithmic Feed for Mastodon that runs 100% in the browser (foryoufeed and fedialgo).

**Comment by abuseofnotation** [comment]
OK, here is the issue that I am having and it will be great if it is solved.

I understand that in order to integrate with instances, I have to first register my application in each instance (https://docs.joinmastodon.org/client/token/#app).

When I do that, I get a client_id which I need when I authorize a user (https://docs.joinmastodon.org/methods/oauth/#authorize)

However, my understanding is that registering the app has to happen once for each instance. In that case, I have to have a server where I would store the authentication data for each instance.

Please, let me know if I made a mistake and/or direct me to the code that does user authentication in the apps that you refer to.

**Comment by pkreissel** [comment]
You could save it but you don't have to. You can just create a new app for each user. right now there is no downside to do it like that. Then you save all the tokens in local storage.

**Comment by abuseofnotation** [comment]
You *can* create a new app, but that's not how the API is designed to be used.

You should not just abuse the API, the correct thing to do is extend it to support this use case, by, for example, have one default empty app that can be used by clients who don't want to register their apps, or cannot in this case.
